src/tools/crawler/metrics.rs

```diff
@@ -4,13 +4,54 @@ use std::{
     fs,
     collections::HashMap,
     time::Duration,
+    net::SocketAddr,
 };
 
 use spectre::{edge::Edge, graph::Graph};
 
 use crate::Crawler;
 
 const LOG_PATH: &str = "crawler-log.txt";
+const LAST_SEEN_CUTOFF: u64 = 10 * 60;
+
+pub struct NetworkMetrics {
+    graph: Graph<SocketAddr>,
+}
+
+impl NetworkMetrics {
+    /// Constructs a new NetworkMetrics with an empty graph.
+    pub fn new() -> NetworkMetrics {
+        NetworkMetrics { graph: Graph::new() }
+    }
+
+    /// Updates the network graph with new connections.
+    pub fn update_graph(&mut self, crawler: &Crawler) {
+        let (drained, connections): (Vec<_>, Vec<_>) = crawler.known_network.connections()
+            .into_iter()
+            .partition(|&conn| {
+                conn.last_seen.elapsed().as_secs() > LAST_SEEN_CUTOFF
+            });
+
+        // Add new connections.
+        for conn in connections {
+            let edge = Edge::new(conn.a, conn.b);
+            if !(self.graph.contains(&edge)) {
```

Comment:
we shouldn't do a lookup here, as the `Edges` are identical (so replacing an existing one with a new one does nothing); it's faster to just do `self.graph.insert(edge);` and not care